types/decimal.go

```diff
@@ -0,0 +1,403 @@
+// valid must come in the form:
+//   (-) whole integers (.) decimal integers
+// examples of acceptable input include:
+//   -123.456
+//   456.7890
+//   345
+//   -456789
+//
+// NOTE an error will return if more decimal places
+// are provided in the string than the constant Precision
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
+		neg = true
+		str = str[1:]
+	}
+
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	strs := strings.Split(str, ".")
+	lenDecs := 0
+	combinedStr := strs[0]
+	if len(strs) == 2 {
+		lenDecs = len(strs[1])
+		if lenDecs == 0 || len(combinedStr) == 0 {
+			return d, ErrUnknownRequest("bad decimal length")
+		}
+		combinedStr = combinedStr + strs[1]
+	} else if len(strs) > 2 {
+		return d, ErrUnknownRequest("too many periods to be a decimal string")
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest("too much Precision in decimal")
+	}
+
+	// add some extra zero's to correct to the Precision factor
+	zerosToAdd := Precision - lenDecs
+	zeros := fmt.Sprintf(`%0`+strconv.Itoa(zerosToAdd)+`s`, "")
+	combinedStr = combinedStr + zeros
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//______________________________________________________________________________________________
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }     // greater than
+func (d Dec) GTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) >= 0 }     // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }    // less than
+func (d Dec) LTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) <= 0 }     // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} } // Is equal to zero
+
+// addition
+func (d Dec) Add(d2 Dec) Dec {
+	res := new(big.Int).Add(d.Int, d2.Int)
+
+	if res.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{res}
+}
+
+// subtraction
+func (d Dec) Sub(d2 Dec) Dec {
+	res := new(big.Int).Sub(d.Int, d2.Int)
+
+	if res.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{res}
+}
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := ChopPrecisionAndRound(mul)
+
+	if chopped.BitLen() > 255+DecimalPrecisionBytes {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
```

Comment:
Do we need two multiplications here? We should be able to do one multiplication and then a pure bankers round. Don't need to block this PR on this though.